keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+
+- `RuntimeClass` API resource definition
+- `RuntimeClass` pod field for specifying the RuntimeClass the pod should be run with
+- Kubelet implementation for fetching & interpreting the RuntimeClass
+- CRI API & implementation for passing along the RuntimeClass parameters
+
+A subsequent update will include the API for describing supported features and an admission
+controller for performing validation & defaulting.
+
+### API
+
+`RuntimeClass` is a new cluster-scoped resource in the `node.k8s.io` API group.
+
+> _The `node.k8s.io` API group would eventually hold the Node resource when `core` is retired.
+> Alternatives considered: `runtime.k8s.io`, `cluster.k8s.io`_
+
+There are 3 categories of fields that may be included in the `RuntimeClassSpec`:
+
+1. **Runtime setup** - These fields control the behavior of the runtime in some way,
+   including configuring arbitrary options in the runtime (`Parameters`), and eventually specifying
+   resource overhead. This is the focus of this initial proposal.
+2. **Compatibility validation** - These fields declare support for various features, and pods are
+   validated against them in admission. This category, along with the admission controller, will be
+   addressed in a subsequent update to this proposal.
+3. **Informational** - These fields surface semantic properties of the runtime to guide runtime
+   selection. Examples could include hardware virtualization, stability, and I/O performance
+   class. These fields are out-of-scope for the initial proposal.
+
+#### API Specification
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+type RuntimeClass struct {
+    metav1.TypeMeta
+    // ObjectMeta minimally includes the RuntimeClass name, which is used to reference the class.
+    // Namespace should be left blank.
+    metav1.ObjectMeta
+
+    Spec RuntimeClassSpec
+}
+
+type RuntimeClassSpec struct {
+    // Parameters holds opaque parameters that are interpreted by the runtime (CRI implementation).
+    Parameters map[string]string
+}
+```
+
+The runtime is selected by the pod by specifying the RuntimeClass in the PodSpec. Once the pod is
+scheduled, the RuntimeClass cannot be changed.
+
+```go
+type PodSpec struct {
+    ...
+    // RuntimeClassName refers to a RuntimeClass object with the same name,
+    // which should be used to run this pod.
+    // +optional
+    RuntimeClassName string
+    ...
+}
+```
+
+The `legacy` RuntimeClass name is reserved. The legacy RuntimeClass is defined to be fully backwards
+compatible with current Kubernetes. This means that the legacy runtime does not specify any runtime
+parameters or perform any feature validation (all features are "supported").
+
+```go
+const (
+    // RuntimeClassNameLegacy is a reserved RuntimeClass name. The legacy
+    // RuntimeClass does not specify any runtime parameters or perform any
+    // feature validation.
+    RuntimeClassNameLegacy = "legacy"
```

Comment:
I'm not a huge fan of the name "legacy".  In my mind, it conveys the idea of "old", "possible a target of deprecation", and "insecure", when really what we are meaning to convey is "the native kubernetes RuntimeClass".

How about "native" instead?

Reply:
> I'm not a huge fan of the name "legacy". In my mind, it conveys the idea of "old", "possible a target of deprecation", and "insecure",

Great, that's exactly what I was going for :D

> How about "native" instead?

I would like to add `native` as a standard RuntimeClass name that users can expect to find on clusters. The difference between `native` and `legacy` is that `native` would include feature validation. This will make more sense once I add the "supported features" update (I went back-and-forth about whether to include it in the initial proposal). But basically `native` will be the runtime that users _should_ use, but `legacy` needs to be included for backwards compatibility. Make sense?

Reply:
By saying that ` is that native would include feature validation.` Do you mean validation on runtime implementation side? Or the following admission design?

Reply:
Following admission design.